### PR TITLE
Fix #4866: do not lift try with no catch block

### LINKED ISSUE #4866: Do not lift try with no catch block
Forward port of scala/scala#922

### PR DESCRIPTION
**Pull request by abeln**
Forward port of https://github.com/scala/scala/pull/922

### CODE REVIEW

File: tests/run/t4866.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+class FooLifted(param: => Object) {
+  val field = try param catch { case e: Exception => () } finally () 
+}
+
+object Test extends App {
+  def numLifted(o: Object): Int = {
+    var lifted = 0
+    
+    for (f <- o.getClass.getDeclaredMethods) {
```
Comment (allanrenucci) [inline tests/run/t4866.scala]:
```scala
def isLifted(method: Method) = method.getName.startsWith("lifted")
o.getClass.getDeclaredMethods.count(isLifted)
```

File: tests/run/t4866.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      if (f.getName.startsWith("lifted")) {
+        lifted += 1
+      }
+    }
+
+    lifted
+  }
+  
+  println("Foo #lifted: %d".format(numLifted(new Foo(new Object))))
+  println("FooLifted #lifted: %d".format(numLifted(new FooLifted(new Object))))
```
Comment (allanrenucci) [inline tests/run/t4866.scala]: Why `format`?
```scala
"Foo #lifted: " + numLifted(new Foo(new Object))
```

File: tests/run/t4866.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// Test that try-finally aren't lifted, but try-catch are.
+
+class Foo(param: => Object) {
```
Comment (allanrenucci) [inline tests/run/t4866.scala]: Do we need a by-name parameter? Let's minimise feature interaction in tests:
```scala
class Foo {
 def foo: Int = try ??? finally ()
}
```

File: tests/run/t4866.scala [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  val field = try param finally ()
+}
+
+class FooLifted(param: => Object) {
```
Comment (allanrenucci) [inline tests/run/t4866.scala]: Same as above

File: compiler/src/dotty/tools/dotc/transform/LiftTry.scala [hunk 5]
```diff
@@ -20,6 +20,12 @@ import util.Store
  *  is lifted to
  *
  *      { def liftedTree$n() = try body catch handler; liftedTree$n() }
+ *
+ * However, don't lift try's without catch expressions (try-finally).
+ * Lifting is needed only for try-catch expressions that are evaluated in a context
+ * where the stack might not be empty. `finally` does not attempt to continue evaluation
+ * after an exception, so the fact that values on the stack are 'lost' does not matter
+ * (copied from https://github.com/scala/scala/pull/922).
```
[the remark this answers is not in the document]
Comment (abeln) [reply compiler/src/dotty/tools/dotc/transform/LiftTry.scala]: Done

File: tests/run/t4866.scala [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Test that try-finally aren't lifted, but try-catch are.
```
[the remark this answers is not in the document]
Comment (abeln) [reply tests/run/t4866.scala]: Done

File: tests/run/t4866.scala [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+class FooLifted(param: => Object) {
+  val field = try param catch { case e: Exception => () } finally () 
+}
+
+object Test extends App {
+  def numLifted(o: Object): Int = {
+    var lifted = 0
+    
+    for (f <- o.getClass.getDeclaredMethods) {
+      f.setAccessible(true)
```
[the remark this answers is not in the document]
Comment (abeln) [reply tests/run/t4866.scala]: Copypasta. Simplified.

File: tests/run/t4866.scala (hunk 1 above)
Comment (abeln) [reply tests/run/t4866.scala]: Thanks. Used it.

File: tests/run/t4866.scala (hunk 2 above)
Comment (abeln) [reply tests/run/t4866.scala]: Removed format.

File: tests/run/t4866.scala (hunk 3 above)
Comment (abeln) [reply tests/run/t4866.scala]: We don't. Removed it.

File: tests/run/t4866.scala (hunk 4 above)
Comment (abeln) [reply tests/run/t4866.scala]: Done.

**Comment by abeln** [comment]
PTAL

**Pull request merged**